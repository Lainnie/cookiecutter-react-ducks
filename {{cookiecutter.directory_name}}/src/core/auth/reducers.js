import { combineReducers } from 'redux';

/**
 * Action Types
 */
const module = 'auth';

export const actionTypes = {
  LOGIN: `${module}/LOGIN`,
  LOGIN_SUCCESS: `${module}/LOGIN_SUCCESS`,
  LOGIN_FAILURE: `${module}/LOGIN_FAILURE`,
  ALREADY_LOGGED_IN: `${module}/ALREADY_LOGGED_IN`,
};

/**
 * Reducers
 */
const loginState = {
  accessToken: '',
  username: '',
  accountId: '',
  error: false,
};

const loginReducer = (state = loginState, action) => {
  switch (action.type) {
    case actionTypes.LOGIN_SUCCESS: {
      const { username, accessToken } = action.payload;

      localStorage.setItem('accessToken', accessToken);
      localStorage.setItem('username', username);

      return {
        ...state,
        accessToken,
        username,
        error: false,
      };
    }
    case actionTypes.ALREADY_LOGGED_IN: {
      const username = localStorage.getItem('username');
      const accessToken = localStorage.getItem('accessToken');

      return {
        ...state,
        accessToken,
        username,
      };
    }
    case actionTypes.LOGIN_FAILURE: {
      return {
        ...loginState,
        error: true,
      };
    }
    default:
      return state;
  }
};

export const initialStates = {
  login: loginState,
};

export const reducers = combineReducers({
  login: loginReducer,
});

/**
 * Operations
 */
const loginSuccess = ({ username, accessToken }) => ({
  type: actionTypes.LOGIN_SUCCESS,
  payload: { username, accessToken },
});

const loginFailure = error => ({
  type: actionTypes.LOGIN_FAILURE,
  payload: { error },
});

const login = (username, password) => ({
  type: actionTypes.LOGIN,
  payload: { username, password },
});

const alreadyLoggedIn = () => ({
  type: actionTypes.ALREADY_LOGGED_IN,
});

export const operations = {
  loginSuccess,
  loginFailure,
  login,
  alreadyLoggedIn,
};

/**
 * Selectors
 */
const getToken = state => state.auth.login.accessToken;
const getUsername = state => state.auth.login.username;
const getAccountId = state => state.auth.login.accountId;
const getSubmitError = state => state.auth.login.error;

export const selectors = {
  getAccountId,
  getUsername,
  getToken,
  getSubmitError,
};
