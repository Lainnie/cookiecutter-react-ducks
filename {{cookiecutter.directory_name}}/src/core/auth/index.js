import { reducers } from './reducers';
import authRemotes from './remotes';

export {
  initialStates as authInitialStates,
  operations as authOperations,
  selectors as authSelectors,
  actionTypes as authActionTypes,
} from './reducers';
export { remotesBuilder as authRemotesBuilder } from './remotes';
export { authRemotes };

export default reducers;
