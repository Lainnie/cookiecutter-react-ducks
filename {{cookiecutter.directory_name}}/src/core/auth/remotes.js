import { call, put, takeLatest } from 'redux-saga/effects';

import { authOperations as _authOperations, authActionTypes } from '../auth';

const LOGIN_ENDPOINT = `${process.env.API_URL}/v1/login`;

const _loginRequest = (fetch, url, options) =>
  fetch(url, options).then(response => response.json());

const remotesBuilder = ({ authOperations, fetch, loginRequest }) => ({
  login: function* login(action) {
    try {
      const headers = {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      };
      const options = {
        method: 'POST',
        body: JSON.stringify({ ...action.payload }),
        headers,
      };
      const response = yield call(loginRequest, fetch, LOGIN_ENDPOINT, options);

      yield put(
        authOperations.loginSuccess({
          username: action.payload.username,
          accessToken: response,
        })
      );
    } catch (e) {
      yield put(authOperations.loginFailure(e));
    }
  },
});

const remotes = remotesBuilder({
  authOperations: _authOperations,
  fetch,
  loginRequest: _loginRequest,
});

export { remotesBuilder };

export default [takeLatest(authActionTypes.LOGIN, remotes.login)];
