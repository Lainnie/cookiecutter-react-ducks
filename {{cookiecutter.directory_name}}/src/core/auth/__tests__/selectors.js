import { authSelectors, authInitialStates } from '../../auth';

describe('auth selectors', () => {
  it('returns error using .getSubmitError', () => {
    const state = {
      auth: {
        ...authInitialStates,
        login: {
          accessToken: '1234567890',
          error: true,
        },
      },
    };

    expect(authSelectors.getSubmitError(state)).toMatchSnapshot();
  });

  it('returns accessToken using .getToken', () => {
    const state = {
      auth: {
        ...authInitialStates,
        login: {
          accessToken: '1234567890',
        },
      },
    };

    expect(authSelectors.getToken(state)).toMatchSnapshot();
  });

  it('returns username using .getUsername', () => {
    const state = {
      auth: {
        ...authInitialStates,
        login: {
          username: 'storetail',
        },
      },
    };

    expect(authSelectors.getUsername(state)).toMatchSnapshot();
  });

  it('returns accountId using .getAccountId', () => {
    const state = {
      auth: {
        ...authInitialStates,
        login: {
          accountId: 1,
        },
      },
    };

    expect(authSelectors.getAccountId(state)).toMatchSnapshot();
  });
});
