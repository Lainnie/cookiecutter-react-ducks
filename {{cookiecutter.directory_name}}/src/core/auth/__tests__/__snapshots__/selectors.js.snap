// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth selectors returns accessToken using .getToken 1`] = `"1234567890"`;

exports[`auth selectors returns accountId using .getAccountId 1`] = `1`;

exports[`auth selectors returns error using .getSubmitError 1`] = `true`;

exports[`auth selectors returns username using .getUsername 1`] = `"storetail"`;
